[Flash161203] iP

Flash161203 opened the pull request:
# Duke

> This chatbot is named after Duke, the Java Mascot.

This chatbot helps you keep track of the tasks you have. It is

- Text-based
- Very simple to learn
- **Super fast** to use


You can also get your own personal task manager up and running in these 3 simple steps:

1. Download v0.1 of the chatbot from [here](https://github.com/Flash161203/ip/releases)
2. Run the jar file
3. Let Duke manage your tasks for you 😉 


### And by the way, it is _completely FREE_!


## The features of Duke are:

- [x] Text-based UI
- [x] Managing to-do's
- [x] Managing deadlines
- [x] Managing events
- [x] Marking tasks as done / undone
- [x] Searching for desired tasks from the list using a keyword
- [ ] Awesome, user-friendly GUI (coming soon)


#### The chatbot, programmed using Java, gets going with a simple function call. The `main` method is as follows:
```java
public static void main(String[] args) {
    Duke bot = new Duke();
    bot.start();

}
```

Commit:
Add Gradle support

Flash161203 committed:
Add Greet, Echo and Exit features

Flash161203 committed:
Minor fixes and UI changes

Flash161203 committed:
Add Listing and Adding features

Flash161203 committed:
Add Tasks, Marking and Unmarking features

Flash161203 committed:
Add ToDo, Deadline and Event features

Flash161203 committed:
Make command name alone case insensitive

Flash161203 committed:
Ignore EXPECTED.TXT

Flash161203 committed:
Revert "Ignore EXPECTED.TXT"

This reverts commit 04ed2d0b0489b995638191f943aca335e449ed01.

Flash161203 committed:
Ignore EXPECTED.TXT

Flash161203 committed:
Add input for Automated Text UI Testing

Flash161203 committed:
Add Exceptions and TaskManager class

Flash161203 committed:
Add Delete feature

Flash161203 committed:
Remove unnecessary imports

Flash161203 committed:
Modularize command processing function

Flash161203 committed:
Add data folder to .gitignore

Flash161203 committed:
Add FileManager class

Flash161203 committed:
Add methods to get file representation of tasks

Flash161203 committed:
Add Save to hard disk feature

Flash161203 committed:
Modify Tasks to store information as date-time

Flash161203 committed:
Add Dates and Times feature

Flash161203 committed:
Merge branch 'branch-Level-7'

Flash161203 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Task.java
#	src/main/java/TaskManager.java

Flash161203 committed:
Fix 'To' date-time before 'From' date-time bug and other minor bugs

Flash161203 committed:
Add Ui class

Flash161203 committed:
Add Storage class

Flash161203 committed:
Add TaskList class to handle operations on the list of tasks

Flash161203 committed:
Add Command class to represent the commands

Flash161203 committed:
Add Parser class to parse user input

Flash161203 committed:
Add DukeInvalidArgumentException in place of DukeEmptyArgumentException

Flash161203 committed:
Delete unrequired files

Flash161203 committed:
Refactor code to use more OOP

Flash161203 committed:
Organize classes into packages and fix minor bugs

Flash161203 committed:
Merge branch 'add-gradle-support' of https://github.com/Flash161203/ip into add-gradle-support

Flash161203 committed:
Merge branch 'add-gradle-support'

Flash161203 committed:
Remove unneccessary import of Task class

Flash161203 committed:
Add unit tests for Deadline and Parser classes

Flash161203 committed:
Change main class name inside build.gradle and fix minor bugs

Flash161203 committed:
Add JavaDoc header comments for all classes in duke.command

Flash161203 committed:
Add JavaDoc header comments for all classes in duke.exception and remove unnecessary imports

Flash161203 committed:
Add JavaDoc header comment for Parser class

Flash161203 committed:
Add JavaDoc header comment for Storage class

Flash161203 committed:
Add JavaDoc header comments for all classes in duke.task

Flash161203 committed:
Add JavaDoc header for TaskList class

Flash161203 committed:
Add JavaDoc header comment for Ui class

Flash161203 committed:
Add JavaDoc header for Duke class

Flash161203 committed:
Tweak code to follow coding standard

Flash161203 committed:
Add Find feature

Flash161203 committed:
Add JavaDoc comments for all public methods of Ui class

Flash161203 committed:
Add JavaDoc comments for execute method in all classes of duke.command

Flash161203 committed:
Add JavaDoc comments for methods in the classes of duke.task

Flash161203 committed:
Add JavaDoc comment for parse method in Parser class

Flash161203 committed:
Add JavaDoc comments for methods in TaskList class

Flash161203 committed:
Fix minor error in JavaDoc for deleteTask method of TaskList class

Flash161203 committed:
Add JavaDoc comments for methods in Storage class

Flash161203 committed:
Merge branch 'branch-A-JavaDoc'

Flash161203 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/exception/DukeInvalidArgumentException.java
#	src/main/java/duke/exception/DukeInvalidCommandException.java

Flash161203 committed:
Merge branch 'branch-Level-9' and add JavaDocs to new classes and methods

# Conflicts:
#	src/main/java/duke/parser/Parser.java

wendy0107 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+        bot.start();
+
+    }
+}
```
LGTM :) Code is easy to read!

wendy0107 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+            }
+
+            if (toExit) {
+                break;
```
Perhaps this if clause could be put above the try-catch?

wendy0107 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,94 @@
+                    LocalDateTime.parse(taskParts[3]),
+                    LocalDateTime.parse(taskParts[4]));
+
+        }
```
Perhaps there should be a final else clause here?

wendy0107 commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,159 @@
+                    + keyword + "' :/");
+
+        } else {
+            System.out.println("Listing all tasks with the keyword '"
```
Perhaps the messages printed in this class could be done in Ui instead!

wendy0107 commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Test
+    public void parse_invalidCommand_exceptionThrown() {
```
Perhaps the exceptions handled here could be done under the exception package instead!

wendy0107 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,274 @@
+     * @return The executable command obtained from the user's input.
+     * @throws DukeException When the user's input cannot be parsed into an executable command.
+     */
+    public static Command parse(String userCommand) throws DukeException {
```
I like how you abstracted the different cases with their own parsing methods!

wendy0107 reviewed (commented):
LGTM! Code is very readable and well-documented. I like how clean your code is, didn't notice any issues either! :D

Flash161203 commented:
> LGTM! Code is very readable and well-documented. I like how clean your code is, didn't notice any issues either! :D

Thank you very much for taking the time to review my code and giving suggestions for improvement. I have taken note of them and will make the necessary changes :)

Flash161203 committed:
Add JavaFX dependencies

Flash161203 committed:
Refactor code to make Ui display responses

Flash161203 committed:
Refactor code for better abstraction

Flash161203 committed:
Add GUI without FXML

Flash161203 committed:
Add GUI to Duke using FXML

Flash161203 committed:
Rename Ui to TextUi

Flash161203 committed:
Merge branch 'branch-Level-10'

Flash161203 committed:
Fix minor bugs

Storage class prints to the output console directly when it encounters
an IOException when writing to the test file.

This is not good behaviour as anything to be printed to the output
consolse must be done by the TextUi class.

So, we will make the Storage object throw a DukeException if it
encounters the IOException, for it to be handled by functions
higher in the call stack.

This will ensure that all possible checked exceptions are
handled at a single place.

Flash161203 committed:
Add assertions and fix minor bugs

Flash161203 committed:
Refactor code to improve code quality

Flash161203 committed:
Merge pull request #2 from Flash161203/branch-A-Assertions

Add assertions and fix minor bugs

Flash161203 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/command/AddCommand.java

Flash161203 committed:
Merge pull request #3 from Flash161203/branch-A-CodeQuality

Improve Code Quality

Flash161203 committed:
Add Priority to Tasks

The tasks cannot be differentiated in terms of their
priorities.

This can reduce the effectiveness of the task
management done by the chatbot.

We will add a Priority attribute to all tasks.

This can help users change the priority
associated with each of the tasks.

Flash161203 committed:
Fix minor bugs and add product screenshot

Flash161203 committed:
Add data file corruption error handling

The program does not take care of the case where the
data file used has been corrupted.

This will lead the program to crash if the data file
is corrupted.

Let's change the program to handle this by making it
delete all data in the file without a warning, and start
afresh, if the file is detected to be corrupted.

This will help avoid the program from crashing in the
event of a data file corruption.

Flash161203 committed:
Make application close on saying bye

Flash161203 committed:
Change find command to display task numbers

Flash161203 committed:
Update README.md

Flash161203 committed:
Update README.md

Flash161203 committed:
Update README.md

Flash161203 committed:
Add better GUI and new product screenshot

Flash161203 committed:
Merge branch 'master' of https://github.com/Flash161203/ip
